integration_test/cases/type.exs

```diff
@@ -91,13 +104,15 @@ defmodule Ecto.Integration.TypeTest do
     assert [^datetime] = TestRepo.all(query)
   end
 
+  @tag :coalesce_type
   test "coalesce type when default" do
```

Comment:
This test also has coalesce:

https://github.com/elixir-ecto/ecto/pull/3223/files#diff-32bcb9b1e35982eb56032061dfab4aa9L378

But it doesn't have a `@tag :coalesce_type`. So do you know why this test fails but that one does not? MSSQL doesn't support coalesce?

Reply:
trick is with types and how ms threat `CAST(value as DECIMAL)` for instance. Decimal without precision is `DECIMAL(X, 0) ` so in case `COALESCE(t1.coumn_name, CAST(1.0 AS DECIMAL))` will not work  since assert expects `#Decimal<1.0>` but in this case we have binary that will preserve all bytes and result is for sure `<<0, 1>>`. 

So coalesce work but CAST is the problem in some cases where precision is required, like in case of Decimal where default precision is 0.

It is worth mentioning here, tds will encode decimal correctly with correct precision, but adapter have no clue about parameter precision or scale used in `select` since `Ecto.Query.Api.type/2`  only pass parameter type that should be used in cast but not other info that is kept in migrations :(

Reply:
I should probably use different tag eg :decimal_casting

Reply:
Both `:decimal_type_cast` and `:db_engine_can_maintain_precision` seem to be related to decimal precision, so what if we have a `@tag :decimal_precision` for all of them?

Do the coalesce tests pass when they don't use decimal?

Reply:
I’m not at my desk but as I can recall, money and decimal types suffer from this in all casting scenarios in select statement